### ISSUE TITLE
tf.reduce_mean gives incorrect results on CPU

### ISSUE THREAD

**Issue by kazimuth** [tensorflow/tensorflow#33373]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Arch Linux
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): v2.0.0-rc2-26-g64c3d38 2.0.0
- Python version: 3.7.4

**Current Behavior**

The following script:

```python
import tensorflow as tf
import numpy as np

(x_train, _), (_, _) = tf.keras.datasets.cifar10.load_data()
x_train = x_train.astype('float32')

tf_mean = tf.reduce_mean(x_train, axis=[0, 1, 2], keepdims=False)
np_mean = np.mean(x_train, (0,1,2))
print('channel means:')
print('  tf:', tf_mean)
print('  np:', np_mean)
```

prints:

```
channel means:
  tf: tf.Tensor([83.88608 83.88608 83.88608], shape=(3,), dtype=float32)
  np: [125.3069  122.95015 113.866  ]
```

Note: the numpy results are the correct channel-wise means for CIFAR10.

**Expected Behavior**
Tensorflow and numpy should give at least vaguely similar results.

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.0  and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/d4946579404e70cca2fbbd56137dfd2d/untitled274.ipynb). Thanks!.

**Comment by rmothukuru** [comment]
The results of `tf.reduce_mean` and `np.mean` are similar if we use `axis = None` or `axis = [0,1,2, 3]`. However, results are very different when `axis = [0,1,2]`. Here is the [Gist](https://colab.sandbox.google.com/gist/rmothukuru/86b2d6db331554600e5c27c10be97831/33373.ipynb#scrollTo=S7CWFytQJi4d).

**Comment by jaingaurav** [comment]
@kazimuth: I believe the problem here is the TensorFlow is doing the computation with 32-bit float values whereas numpy is using 64-bit floating almost all the time. Even if you change `np.mean` to use `dtype=float32`, it still seems to perform the calculation with 64-bit float values. If you want the same behavior I suggest you change x_train to use float64 and then cast to float32 in the very end like so:

```
import tensorflow as tf
import numpy as np

(x_train, _), (_, _) = tf.keras.datasets.cifar10.load_data()
x_train = x_train.astype('float64')

tf_mean = tf.cast(tf.reduce_mean(x_train, axis=[0, 1, 2,], keepdims=False), dtype=tf.float32)
np_mean = np.mean(x_train, (0,1,2), dtype=np.float32)
print('channel means:')
print('  tf:', tf_mean)
print('  np:', np_mean)
```
